rasa/utils/tensorflow/layers_utils.py

```diff
@@ -71,14 +71,24 @@ def get_candidate_values(
     return candidate_values  # (batch_size, num_candidates, tf.shape(x)[-1])
 
 
-def reduce_mean_equal(x: tf.Tensor, y: tf.Tensor) -> tf.Tensor:
+def reduce_mean_equal(
+    x: tf.Tensor, y: tf.Tensor, mask: Optional[tf.Tensor] = None
+) -> tf.Tensor:
     """Computes the mean number of matches between x and y.
 
     Args:
-        x: Any numeric tensor
-        y: Another tensor with same shape and type as x
+        x: Any numeric tensor.
+        y: Another tensor with same shape and type as x.
+        mask: Tensor with a mask to distinguish actual indices from padding indices.
+            Shape should be the same as `x` and `y`.
 
     Returns:
         The mean of "x == y"
     """
-    return tf.reduce_mean(tf.cast(tf.math.equal(x, y), tf.float32))
+    if mask is None:
+        return tf.reduce_mean(tf.cast(tf.math.equal(x, y), tf.float32))
+    else:
+        equal_indices = tf.cast(tf.math.equal(x, y), tf.float32) * mask
+        return tf.reduce_mean(
+            tf.reduce_sum(equal_indices, axis=-1) / tf.reduce_sum(mask, axis=-1)
```

Comment:
As advertised in the docstring and name, this function seems agnostic to the dimensions, yet when a mask is passed in it suddenly treats the last dimension as the one of concern and makes everything else a batch dimension over which it then averages. This gives counter intuitive behavior. Based on the behavior of the unmasked case, I would expect
```python
x = [[1,2,3,4]
     [5,6,7,8]] 
y = [[1,2,3,4]
     [5,6,0,0]]
mask = [[1,1,1,1],
        [1,1,1,0]]
```
to give `6 / 7` since there are 6 total matches on 7 unmasked items.
What is currently computed however is  `((4/4) + 2/3) / 2`
             
Perhaps we should add an axis argument that gets passed so you can specify the desired behavior explicitly similar to the underlying tensorflow ops `reduce_mean` and `reduce_sum` upon which this operation is based?

Reply:
I clarified it in the docstrings for now. We should give it a thought on how to make it more general purpose in the future (adding the axis argument).

Reply:
I'll create an issue for it.